File: app/src/main/java/de/xikolo/models/base/Sync.java
package de.xikolo.models.base;

import android.util.Log;
import android.util.Pair;

import java.util.ArrayList;
import java.util.List;

import de.xikolo.config.Config;
import io.realm.Realm;
import io.realm.RealmModel;
import io.realm.RealmQuery;
import io.realm.RealmResults;
import moe.banana.jsonapi2.Document;
import moe.banana.jsonapi2.Resource;

public abstract class Sync<S extends RealmModel, T extends Resource & RealmAdapter<S>> {

    public static final String TAG = Sync.class.getSimpleName();

    Class<S> clazz;

    List<Pair<String, String>> filters;

    List<Pair<String, String[]>> inFilters;

    BeforeCommitCallback<S> beforeCommitCallback;

    boolean handleDeletes;

    Sync(Class<S> clazz) {
        this.clazz = clazz;
        this.filters = new ArrayList<>();
        this.inFilters = new ArrayList<>();
        this.handleDeletes = true;
    }

    public Sync<S, T> addFilter(String fieldName, String value) {
        this.filters.add(new Pair<>(fieldName, value));
        return this;
    }

    public Sync<S, T> addFilter(String fieldName, String[] values) {
        this.inFilters.add(new Pair<>(fieldName, values));
        return this;
    }

    public Sync<S, T> setBeforeCommitCallback(BeforeCommitCallback<S> callback) {
        this.beforeCommitCallback = callback;
        return this;
    }

    /**
     * Sync resources without deleting untouched resources
     */
    public Sync<S, T> saveOnly() {
        this.handleDeletes = false;
        return this;
    }

    public interface BeforeCommitCallback<S extends RealmModel>  {
        void beforeCommit(Realm realm, S model);
    }

    public abstract String[] run();

    public static class Data<S extends RealmModel, T extends Resource & RealmAdapter<S>> extends Sync<S, T> {

        private T[] items;

        private Data(Class<S> clazz, final T[] items) {
            super(clazz);
            this.items = items;
        }

        @SafeVarargs
        public static <S extends RealmModel, T extends Resource & RealmAdapter<S>> Data<S, T> with(Class<S> clazz, final T... items) {
            return new Data<>(clazz, items);
        }

        @Override
        public String[] run() {
            final List<String> ids = new ArrayList<>();

            try (Realm realmInstance = Realm.getDefaultInstance()) {
                realmInstance.executeTransaction(realm -> {
                    for (T item : items) {
                        S model = item.convertToRealmObject();
                        if (beforeCommitCallback != null) {
                            beforeCommitCallback.beforeCommit(realm, model);
                        }
                        realm.copyToRealmOrUpdate(model);
                        ids.add(item.getId());
                    }

                    if (Config.DEBUG) Log.d(TAG, "DATA: Saved " + items.length + " data resources from type " + clazz.getSimpleName());

                    if (handleDeletes) {
                        RealmQuery<S> deleteQuery = realm.where(clazz);
                        if (ids.size() > 0) {
                            deleteQuery.not().in("id", ids.toArray(new String[0]));
                        }
                        if (filters != null) {
                            for (Pair<String, String> filter : filters) {
                                deleteQuery.equalTo(filter.first, filter.second);
                            }
                        }
                        if (inFilters != null) {
                            for (Pair<String, String[]> filter : inFilters) {
                                deleteQuery.in(filter.first, filter.second);
                            }
                        }
                        RealmResults<S> results = deleteQuery.findAll();

                        if (Config.DEBUG) Log.d(TAG, "DATA: Deleted " + results.size() + " local resources from type " + clazz.getSimpleName());

                        results.deleteAllFromRealm();
                    } else if (Config.DEBUG) Log.d(TAG, "DATA: Deleted 0 local resources from type " + clazz.getSimpleName());
                });
            }

            return ids.toArray(new String[0]);
        }

    }

    public static class Included<S extends RealmModel, T extends Resource & RealmAdapter<S>> extends Sync<S, T> {

        private Document document;

        private Included(Class<S> clazz, Document document) {
            super(clazz);
            this.document = document;
        }

        public static <S extends RealmModel> Included<S, ?> with(Class<S> clazz, final Document document) {
            return new Included<>(clazz, document);
        }

        public static <S extends RealmModel> Included<S, ?> with(Class<S> clazz, final Resource... items) {
            if (items.length > 0) {
                return new Included<>(clazz, items[0].getDocument());
            }
            return new Included<>(clazz, null);
        }

        @Override
        public String[] run() {
            if (document == null) return new String[0];

            final List<String> ids = new ArrayList<>();

            try (Realm realmInstance = Realm.getDefaultInstance()) {
                realmInstance.executeTransaction(realm -> {
                    for (Resource resource : document.getIncluded()) {
                        if (resource instanceof RealmAdapter) {
                            RealmAdapter adapter = (RealmAdapter) resource;
                            RealmModel model = adapter.convertToRealmObject();
                            if (model.getClass() == clazz) {
                                if (beforeCommitCallback != null) {
                                    //noinspection unchecked
                                    beforeCommitCallback.beforeCommit(realm, (S) model);
                                }
                                realm.copyToRealmOrUpdate(model);
                                ids.add(resource.getId());
                            }
                        }
                    }

                    if (Config.DEBUG) Log.d(TAG, "INCLUDED: Saved " + ids.size() + " included resources from type " + clazz.getSimpleName());

                    if (handleDeletes) {
                        RealmQuery<S> deleteQuery = realm.where(clazz);
                        if (ids.size() > 0) {
                            deleteQuery.not().in("id", ids.toArray(new String[0]));
                        }
                        if (filters != null) {
                            for (Pair<String, String> filter : filters) {
                                deleteQuery.equalTo(filter.first, filter.second);
                            }
                        }
                        if (inFilters != null) {
                            for (Pair<String, String[]> filter : inFilters) {
                                deleteQuery.in(filter.first, filter.second);
                            }
                        }
                        RealmResults<S> results = deleteQuery.findAll();

                        if (Config.DEBUG) Log.d(TAG, "INCLUDED: Deleted " + results.size() + " local resources from type " + clazz.getSimpleName());

                        results.deleteAllFromRealm();
                    } else if (Config.DEBUG) Log.d(TAG, "INCLUDED: Deleted 0 local resources from type " + clazz.getSimpleName());
                });
            }

            return ids.toArray(new String[0]);
        }

    }

}

File: app/src/main/java/de/xikolo/network/sync/Sync.kt
package de.xikolo.network.sync

import android.util.Log
import de.xikolo.config.Config
import de.xikolo.models.base.RealmAdapter
import io.realm.Realm
import io.realm.RealmModel
import io.realm.kotlin.oneOf
import moe.banana.jsonapi2.Document
import moe.banana.jsonapi2.Resource
import java.util.*

abstract class Sync<S : RealmModel> internal constructor(internal var clazz: Class<S>) {

    companion object {
        val TAG: String = Sync::class.java.simpleName
    }

    internal var filters: MutableList<Pair<String, String>> = ArrayList()

    internal var inFilters: MutableList<Pair<String, Array<out String>>> = ArrayList()

    internal var beforeCommitCallback: ((realm: Realm, model: S) -> Unit)? = null

    internal var handleDeletes: Boolean = true

    fun addFilter(fieldName: String, value: String): Sync<S> {
        this.filters.add(Pair(fieldName, value))
        return this
    }

    fun addFilter(fieldName: String, values: Array<out String>): Sync<S> {
        this.inFilters.add(Pair(fieldName, values))
        return this
    }

    fun setBeforeCommitCallback(callback: (realm: Realm, model: S) -> Unit): Sync<S> {
        this.beforeCommitCallback = callback
        return this
    }

    /**
     * Sync resources without deleting untouched resources
     */
    fun saveOnly(): Sync<S> {
        this.handleDeletes = false
        return this
    }

    abstract fun run(): Array<out String>

    class Data<S : RealmModel, T> private constructor(clazz: Class<S>, private val items: Array<T>) : Sync<S>(clazz)
            where T : Resource,
                  T : RealmAdapter<S> {

        companion object {

            fun <S : RealmModel, T> with(clazz: Class<S>, items: Array<T>): Data<S, T>
                    where T : Resource,
                          T : RealmAdapter<S> {
                return Data(clazz, items)
            }

            inline fun <reified S : RealmModel, reified T> with(items: Array<T>): Data<S, T>
                    where T : Resource,
                          T : RealmAdapter<S> {
                return with(S::class.java, items)
            }

            inline fun <reified S : RealmModel, reified T> with(item: T): Data<S, T>
                    where T : Resource,
                          T : RealmAdapter<S> {
                return with(S::class.java, arrayOf(item))
            }

        }

        override fun run(): Array<out String> {
            val ids = ArrayList<String>()

            Realm.getDefaultInstance().use { realmInstance ->
                realmInstance.executeTransaction { realm ->
                    for (item in items) {
                        val model = item.convertToRealmObject()
                        beforeCommitCallback?.invoke(realm, model)
                        realm.copyToRealmOrUpdate(model)
                        ids.add(item.id)
                    }

                    if (Config.DEBUG) Log.d(TAG, "DATA: Saved ${items.size} data resources from type ${clazz.simpleName}")

                    if (handleDeletes) {
                        val deleteQuery = realm.where(clazz)
                        if (ids.size > 0) {
                            deleteQuery.not().oneOf("id", ids.toTypedArray())
                        }
                        for (filter in filters) {
                            deleteQuery.equalTo(filter.first, filter.second)
                        }
                        for (filter in inFilters) {
                            deleteQuery.oneOf(filter.first, filter.second)
                        }
                        val results = deleteQuery.findAll()

                        if (Config.DEBUG) Log.d(TAG, "DATA: Deleted ${results.size} local resources from type ${clazz.simpleName}")

                        results.deleteAllFromRealm()
                    } else if (Config.DEBUG) Log.d(TAG, "DATA: Deleted 0 local resources from type ${clazz.simpleName}")
                }
            }

            return ids.toTypedArray()
        }

    }

    class Included<S : RealmModel> private constructor(clazz: Class<S>, private val document: Document?) : Sync<S>(clazz) {

        companion object {

            fun <S : RealmModel> with(clazz: Class<S>, items: Array<out Resource>): Included<S> {
                return if (items.isNotEmpty()) {
                    Included(clazz, items.first().document)
                } else {
                    Included(clazz, null)
                }
            }

            inline fun <reified S : RealmModel> with(items: Array<out Resource>): Included<S> {
                return with(S::class.java, items)
            }

            inline fun <reified S : RealmModel> with(item: Resource): Included<S> {
                return with(S::class.java, arrayOf(item))
            }

        }

        override fun run(): Array<out String> {
            if (document == null) return arrayOf()

            val ids = ArrayList<String>()

            Realm.getDefaultInstance().use { realmInstance ->
                realmInstance.executeTransaction { realm ->
                    for (resource in document.included) {
                        if (resource is RealmAdapter<*>) {
                            val adapter = resource as RealmAdapter<*>
                            val model = adapter.convertToRealmObject()
                            if (model.javaClass == clazz) {
                                @Suppress("UNCHECKED_CAST")
                                beforeCommitCallback?.invoke(realm, model as S)
                                realm.copyToRealmOrUpdate(model)
                                ids.add(resource.id)
                            }
                        }
                    }

                    if (Config.DEBUG) Log.d(TAG, "INCLUDED: Saved ${ids.size} included resources from type ${clazz.simpleName}")

                    if (handleDeletes) {
                        val deleteQuery = realm.where(clazz)
                        if (ids.size > 0) {
                            deleteQuery.not().oneOf("id", ids.toTypedArray())
                        }
                        for (filter in filters) {
                            deleteQuery.equalTo(filter.first, filter.second)
                        }
                        for (filter in inFilters) {
                            deleteQuery.oneOf(filter.first, filter.second)
                        }
                        val results = deleteQuery.findAll()

                        if (Config.DEBUG) Log.d(TAG, "INCLUDED: Deleted ${results.size} local resources from type ${clazz.simpleName}")

                        results.deleteAllFromRealm()
                    } else if (Config.DEBUG) Log.d(TAG, "INCLUDED: Deleted 0 local resources from type ${clazz.simpleName}")
                }
            }

            return ids.toTypedArray()
        }

    }

}
